warp_drive/training/lightning_trainer.py

```diff
@@ -0,0 +1,996 @@
+            if self.verbose:
+                verbose_print(f"Saving the results to the file '{results_filename}'")
+            with open(results_filename, "a+", encoding="utf8") as fp:
+                json.dump(metrics, fp)
+                fp.write("\n")
+
+    def load_model_checkpoint(self, ckpts_dict=None):
+        """
+        Load the model parameters if a checkpoint path is specified.
+        """
+        if ckpts_dict is None:
+            logging.info(
+                "Loading trainer model checkpoints from the run configuration."
+            )
+            for policy in self.policies:
+                ckpt_filepath = self.config["policy"][policy]["model"][
+                    "model_ckpt_filepath"
+                ]
+                self._load_model_checkpoint_helper(policy, ckpt_filepath)
+        else:
+            assert isinstance(ckpts_dict, dict)
+            if self.verbose:
+                verbose_print("Loading the provided trainer model checkpoints.")
+            for policy, ckpt_filepath in ckpts_dict.items():
+                assert policy in self.policies
+                self._load_model_checkpoint_helper(policy, ckpt_filepath)
+
+    def _load_model_checkpoint_helper(self, policy, ckpt_filepath):
+        if ckpt_filepath != "":
+            assert os.path.isfile(ckpt_filepath), "Invalid model checkpoint path!"
+            if self.verbose:
+                verbose_print(
+                    f"Loading the '{policy}' torch model "
+                    f"from the previously saved checkpoint: '{ckpt_filepath}'"
+                )
+            self.models[policy].load_state_dict(torch.load(ckpt_filepath))
+
+            # Update the current timestep using the saved checkpoint filename
+            timestep = int(ckpt_filepath.split(".state_dict")[0].split("_")[-1])
+            if self.verbose:
+                verbose_print(
+                    f"Updating the timestep for the '{policy}' model to {timestep}.",
+                )
+            self.current_timestep[policy] = timestep
+
+    def save_model_checkpoint(self, iteration=0):
+        """
+        Save the model parameters
+        """
+        # Save model checkpoints if specified (and also for the last iteration)
+        if (
+            iteration % self.config["saving"]["model_params_save_freq"] == 0
+            or iteration == self.num_iters - 1
+        ):
+            for policy, model in self.models.items():
+                filepath = os.path.join(
+                    self.save_dir,
+                    f"{policy}_{self.current_timestep[policy]}.state_dict",
+                )
+                if self.verbose:
+                    verbose_print(
+                        f"Saving the '{policy}' torch model "
+                        f"to the file: '{filepath}'."
+                    )
+
+                torch.save(model.state_dict(), filepath)
+
+    def graceful_close(self):
```

Comment:
say what happens if you don't close gracefully

Reply:
The Cuda backend has dynamic memory allocated for random sampling seeds, those dynamic memory is at the memory heap and needs to be properly deleted before the program exits, otherwise it will have a memory leakage. In warpdrive, we have Python frontend control the CUDA backend to do this clean-up before everything is finished.